### ISSUE TITLE
Separate binstubs directory from bin builds directory

### ISSUE THREAD

**Issue by christhekeele** [bundler/bundler-features#105]
Currently, `bundle binstubs` installs binstubs in the root `bin/` folder of the project.

However, `bundle gem` creates scaffolding that assumes that you're creating a gem that will place its own executable bin files in the same `bin/`.

This conflict culminates in the default gemspec which reads, whether or not you supplied the `--bin` options to `bundle gem`,

```ruby
Gem::Specification.new do |spec|
  #...
  spec.executables   = spec.files.grep(%r{^bin/}) { |f| File.basename(f) }
  #...
end
```

Therefore the default bundler behavior unintentionally encourages you to package your binstubs with your project's executables and distribute them, which is bad. 

Binstubs are normally not a part of library projects, and custom executables are not normally a part of application projects, so this occurs rarely, but it has happened to me several times when building complicated gems with custom binstubs. My proper bins have been stomped on by my own library's binstubs when I run `rake install` from `bundler/gem_tasks` to take my executable out for a spin. I've even released to my own gem repository and broke a few builds this way before.

Changing the default line in the gemspec, or making sure to add every binstub to your `.gitignore` file fixes this easily, and is even something bundler could do for you, but I think this points to a conceptual flaw: bundler groups two different types of bin file into the same location.

I'd like to propose placing project executables in `bin/` and binstubs in `bin/stubs/`, and updating the default `spec.executables` and `.gitignore` to avoid the now-namespaced `/bin/stubs` folder entirely.

I doubt this would be a teeny-level change but it seems a little saner to me.

This isn't exactly a bug, nor a feature request, just unexpected behaviour; so let me know if there's a better place to discuss this proposal or a valid reason to totally dismiss it.

**Comment by segiddins** [comment]
We actually place project executables in `exe` now, perhaps we should update the `bundle gem` output to do the same, @indirect ?

**Comment by christhekeele** [comment]
I'm planning on doing a hack night this evening, so I can throw together a PR tonight.

**Comment by segiddins** [comment]
Excellent, thanks @christhekeele !

**Comment by christhekeele** [comment]
On review, it looks like this has already hit `master`, and possibly has even been released (haven't updated in a spell).

The only existing vestige of the old use of the `bin` dir for executables I can find in `master` is the actual flag supplied to `bundle gem`, which still uses `--bin` and `-b` I assume for backwards compatibility.

I can make a PR to support `--exe`, and use that option name as the canonical reference in the code, but still also check `--bin` and `-b` to get its value; and a second that builds on top of it to totally deprecate the old flags; both of which can be merged at whatever point is appropriate in bundler's release cycle. Aside from that, there's not much else to do.

I'll also give the documentation at `bundler/bundler-site` a pass and see if a) it needs to be updated anywhere to reflect the change to `exe/`, and b) submit separate PRs reflecting the changes to the flags, to be merged in concurrently with their implementation.

Sound good?

**Comment by segiddins** [comment]
Sounds great!

**Comment by christhekeele** [comment]
Turns out that this has actually been out for about 9 months, but I generated scaffolding for my latest gem that long ago. `bundle gem` is kind of how I bookmark new project ideas, I guess.

I've submitted requests to roll out deprecation of the old flag, updating the manfiles accordingly.

`bundler/bundler-site` doesn't have a clear area for adding unreleased documentation for upcoming versions, so I've left it alone.

Thanks for being quick on the response and letting me know about the `exe` change! I hope these little PRs can represent a bit of my gratitude for all bundler team does.
